torch/csrc/jit/script/exit_transforms.cpp

```diff
@@ -0,0 +1,394 @@
+  }
+
+  ExitStatus getStatus(ExitPair& exit_pair) {
+    Value* exit_v = exit_pair.hasExited();
+    if (exit_v == true_val) {
+      return WILL;
+    } else if (exit_v == false_val) {
+      return WONT;
+    } else {
+      return MIGHT;
+    }
+  }
+
+  // Guards the remaining nodes in the block with an if node that takes
+  // the has exited value as its conditional
+  ExitPair guardBlockNodes(
+      Block* block,
+      const ExitPair& exit_pair,
+      graph_node_list_iterator& iter) {
+    auto new_if = graph->create(prim::If, 0)->insertBefore(*iter);
+    new_if->addInput(exit_pair.hasExited());
+
+    auto exit_block = new_if->addBlock();
+    auto guard_block = new_if->addBlock();
+
+    // Move all remaining nodes into the guard block
+    while (iter != block->nodes().end()) {
+      auto node = *iter++;
+      node->moveBefore(guard_block->return_node());
+    }
+
+    std::vector<Value*> exit_block_vals;
+    {
+      // after an exit, the only values that will get used
+      // are the hasExited() and exitValues(), so we match the existing
+      // block outputs with unitialized
+      WithInsertPoint insert(exit_block);
+      exit_block_vals = matchValuesWithUnitialized(block->outputs());
+    }
+
+    // Set the new if to have the same outputs of the original block,
+    // then replace the original block outputs with new if's outputs
+    for (size_t i = 0; i < block->outputs().size(); ++i) {
+      exit_block->registerOutput(exit_block_vals.at(i));
+      guard_block->registerOutput(block->outputs().at(i));
+      new_if->addOutput()->setType(block->outputs().at(i)->type());
+    }
+
+    while (block->outputs().size() > 0) {
+      block->eraseOutput(0);
+    }
+    for (auto out : new_if->outputs()) {
+      block->registerOutput(out);
+    }
+
+    graph->create(prim::LoopExit, {exit_pair.exitValues()}, 0)
+        ->insertBefore(exit_block->return_node());
+    return handleIf(new_if);
+  }
+
+  // these nodes my have uses,
+  // such as in the case:
+  // if i == 1:
+  //    break
+  //    j = j + 1
+  // where the j + 1 value will be a block output, but since they will
+  // never be used, it is safe to replace them with unitialized value
+  void destroyNodeAfterExit(Node* n) {
+    for (auto output : n->outputs()) {
+      if (output->uses().size() > 0) {
+        output->replaceAllUsesWith(getUnitValue(output->type()));
+      }
+    }
+    n->destroy();
+  }
+
+  void deleteAfterExitNodes(Block* block, graph_node_list_iterator& iter) {
+    if (iter == block->nodes().end()) {
+      return;
+    }
+    WithInsertPoint insert(*block->nodes().begin());
+    // need to destroy in reverse order so nodes have no uses when destroyed
+    for (auto it = block->nodes().reverse().begin(); it != iter;) {
+      Node* n = *it++;
+      if (*it != block->return_node()) {
+        destroyNodeAfterExit(n);
+      }
+    }
+    destroyNodeAfterExit(*iter);
+  }
+
+  ExitPair transformExits(Block* block) {
```

Comment:
I'm not sure that restructuring it form a taking a Block to Node will be all that helpful. When we restructure returns for Graphs & Closures, we have to operate on the block b/c a Graph is not a node.  

For the second part: 1) figuring out if the node can break/continue, 2) transforming the rest of the block correspondingly it seems like we'd just be duplicating work without all that much benefit. The current pass already inserts the minimal number of outputs as it is computed, so it's not like calculating ahead of time will really gain us much. 
That is: we don't add any outputs if it WONT_EXIT, and we add exactly the outputs we need otherwise (no extra boolean output if it is WILL)

Reply:
> I'm not sure that restructuring it form a taking a Block to Node will be all that helpful. When we restructure returns for Graphs & Closures, we have to operate on the block b/c a Graph is not a node.

From my experience it just helps to "set the stage" - some parts of the code clearly should not look further than the current node, some functions shouldn't look deeper than the current block. But that's not a hard rule, so you might be right that in this case it doesn't improve anything.

> For the second part: 1) figuring out if the node can break/continue, 2) transforming the rest of the block correspondingly it seems like we'd just be duplicating work without all that much benefit.

Here as well you might be right, I'm just suggesting random ideas that could make the code easier to read.

> The current pass already inserts the minimal number of outputs as it is computed, so it's not like calculating ahead of time will really gain us much.
That is: we don't add any outputs if it WONT_EXIT, and we add exactly the outputs we need otherwise (no extra boolean output if it is WILL)

Is it actually justified? I'd rather have this one extra argument that is in some cases constant true, but then I would know that the first argument is always the `has_exited` independently on the context the node came from and the code of this function would be smaller.

Reply:
The main reason I took the current approach is because it means `exit_pair` doesn't needs to have another field. This was a comment @zdevito had made in a prior version. We can either maintain the invariant that `to_constant<bool>` either equals whether it will exit, or if it's not a constant than it might exit, or we can add another field to exit_pair. The extra if outputs don't particularly matter.

Small change either way, let me know your preference and i will do it.